### PR TITLE
Investigations in the initialization of coq binaries and command line (part 1)

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** architecture of command line and coq binaries initialization

This is the first part of a split of #10147 into smaller reviewable components. If this is too many commits, I can split even further.

I copy below the relevant part of #10147.

---
- Issues mentioned in #9585:

  - It tries to clarify the three roles of `coqtop` (in `coqargs.ml`)
     - interactive mode: we initialize and run the read-eval-loop
     - batch mode: we interpret the command line and quit
     - query mode (`-config`, `-tags`, `-where`, `-print-mod-ui`, `--version`, `--help`): we report about some data and nothing else

    Note that I don't see this splitting as absolutely necessary. I see it as a way to help us structuring options of the command line better. So, it is really more about internal design only.

  - It tries to make more apparent the phase 1 (general initialization) / phase 2 (configuration) / phase 3 (environment loading) of the initialization of a coq process (it seems to me to make understanding of the flow easier to read, but this is somehow also subjective).

- Extension of the parametrization of `Coqtop.init_toplevel` via a set of `custom` methods specific to each binary:
---

Note the discussion in #9585: I'm more and more considering that it would make more sense to see the command line as an ordered sequence of vernac commands than checking what combinations or repetitions in the command line are meaningful. However, this PR does not step in this direction.

### CODE REVIEW

**Comment by ejgallego** [comment]
Size seems OK to me @herbelin , I'll try to review ASAP.

File: toplevel/coqtop.ml
Comment (ejgallego) [inline toplevel/coqtop.ml]: It seems `print_memory_stat` is not called from anywhere?

File: toplevel/coqtop.ml
Comment (ejgallego) [inline toplevel/coqtop.ml]: This is better named as `init_document`, the stm has its own initialization there.

File: toplevel/coqtop.ml
Comment (ejgallego) [inline toplevel/coqtop.ml]: Does this typecheck?

File: ide/idetop.ml
Comment (ejgallego) [inline ide/idetop.ml]: Why this change?

**Comment by ejgallego** [comment]
Looks pretty good I think @herbelin , a couple of questions do remain + the failures.

I am also not sure about the last commit, maybe it should be spun off to a different PR ?

The changes in workers need review from @gares 

IMHO, once @gares reviews and the failures are gone we could merge, tho I'd like to have a deeper look to the last commit so it may be quicker if we just process the rest while I can study that last one.

My reason is that as you know I have a few other toplevels [jsCoq / serAPI] and I'd like to understand how the new API works with them, but that will take me a bit of time.

File: toplevel/coqtop.ml
Comment (herbelin) [reply toplevel/coqtop.ml]: I moved it to `init_execution` (triggered by option `-m`)

Comment (ejgallego) [reply toplevel/coqtop.ml]: Umm, I cannot see it in https://github.com/coq/coq/pull/10246/commits/b2a3df4b81bb93302968fc07c6d4fa77e94a5933

Comment (herbelin) [reply toplevel/coqtop.ml]: Indeed, I mixed up lines, probably while rebasing, and the corresponding was in another commit. I fixed it.

File: toplevel/coqtop.ml
Comment (herbelin) [reply toplevel/coqtop.ml]: Done.

File: toplevel/coqtop.ml
Comment (herbelin) [reply toplevel/coqtop.ml]: I think so (received by the `let state, opts = ...`).

File: toplevel/coqtop.ml
Comment (ejgallego) [reply toplevel/coqtop.ml]: Thanks!

File: ide/idetop.ml
Comment (herbelin) [reply ide/idetop.ml]: The idea was to remove `-batch` from the common command line options parser to the executables for which it really matters. In particular, I could not see a meaning for `-batch` for `coqc`. So I dispatched the test for `Batch` vs `Interactive` mode in `coqtop` and `coqidetop`: `coqidetop` is still called both interactively (to communicate really) and batch (to check connection).

What did you have in mind?

Comment (ejgallego) [reply ide/idetop.ml]: Thanks for the explanation, I understand now; I just didn't get the full picture.

**Comment by gares** [comment]
I need more time to review the priority thing

File: stm/asyncTaskQueue.mli
Comment (gares) [inline stm/asyncTaskQueue.mli]: comments are outdated

File: toplevel/workerLoop.ml
Comment (gares) [inline toplevel/workerLoop.ml]: @ejgallego this question is for you

File: toplevel/workerLoop.mli
Comment (gares) [inline toplevel/workerLoop.mli]: I don't like this string argument, since it may get out of sync with `T.name` (see AsyncTaskQueue.Make argument).
IMO, you should get the name out of `Sys.executable_name` or something like that.

**Review by gares — COMMENTED** [review]
The parts involving the STM are OKish, I've commented in a few places were improvement could be made.

File: toplevel/workerLoop.mli
Comment (herbelin) [reply toplevel/workerLoop.mli]: I made this by consistency with `coqtop` and `coqc` whose name for `--help` is similarly hard-wired in the usage text. We could indeed alternatively use `Sys.executable_name` in each of the cases. (For instance, I tested with `ocamlc` and it uses `Sys.executable_name`. I tested with `gcc` and it uses a hard-wired name, from what I conclude that there is somehow an arbitrary choice.)

If no further comments about that in the discussion, I shall use `Sys.executable_name`.

File: stm/asyncTaskQueue.mli
Comment (herbelin) [reply stm/asyncTaskQueue.mli]: Thanks, I overlooked it. I added `[pri]` in the description of the function.

About this, I have actually a more general question. Why not to pass all of `cur_opt : stm_opt` functionally in the stm. It seems to me that this would make reasoning on the code simpler, showing more explicitly when it is modified or not. (I suspect that it is in fact not modified.)

Also, I found some references to `-toploop` here and there (precisely in `async_proofs_is_master` and `fakeide.ml`) and I guess that it is obsolete comment. Can I remove them?

Even more generally, one of my motivation was to understand if we could simplify the massaging of command line option names at the level of strings in `AsyncTaskQueue.spawn`. At some time I thought we could send the whole `Coqargs.t` directly over the channel as the initial `opts` field of the coqworkers but this would require the type `Coqargs.t` to be known already from the `stm`, i.e. to typically define it in `vernac`. Would it be too demanding? I don't see otherwise how we could pass the command line state to the coq worker (somehow, massaging the string in `asyncTaskQueue.ml` is an evidence that something of the command line argument structure has to be known at the level of the stm!)

**Comment by ejgallego** [comment]
Please rebase so we see if the CI failures are still there.

**Comment by ejgallego** [comment]
Yup, this seems still broken; I'd suggest to try with the first commits that seems safe so we can merge that part quicker, whlie the other commits are fixed.

File: ide/idetop.ml
Comment (ejgallego) [reply ide/idetop.ml]: Maybe you would like to write a more descriptive commit name tho, "fix" seems too terse IMVHO.

Comment (herbelin) [reply ide/idetop.ml]: Sorry, I intended to squash it with the commit just before then forgot. Thanks for noticing. I now squashed it.

Comment (ejgallego) [reply ide/idetop.ml]: Thanks @herbelin , will try to have a look and merge ASAP.

File: toplevel/workerLoop.ml
Comment (ejgallego) [reply toplevel/workerLoop.ml]: The workers don't use the STM API that requires a doc handler.

**Review by ejgallego — APPROVED** [review]
Looks good to me, I propose we merge in a couple of days in case @gares has some more comments.

There is still work to do here, in particular w.r.t. initialization functions but that shall fit well after this PR is merged.

Note the question I ask below.

File: ide/idetop.ml
Comment (ejgallego) [inline ide/idetop.ml]: @herbelin can you add a comment / clarify how stdout is involved here? It is not obvious to me what is going on, but I am not expert on IDE code.

Using stdout in the parse function seems a bit strange to me.

Comment (ejgallego) [reply ide/idetop.ml]: Ping @herbelin

Comment (herbelin) [reply ide/idetop.ml]: I'll look at that within a few days.

Comment (herbelin) [reply ide/idetop.ml]: @ejgallego: I added the following comments:
```
+  (* One of the role of coqidetop is to find the name of buffers to open *)
+  (* in the command line; Coqide is waiting these names on stdout *)
+  (* (see filter_coq_opts in coq.ml), so we send them now *)
```
I'm not saying that this is a good way to do. Maybe the names of buffers to open should be passed instead on a special communication channel?

Comment (ejgallego) [reply ide/idetop.ml]: Thanks for the info @herbelin , let me have a look.

Comment (ejgallego) [reply ide/idetop.ml]: Indeed I understand the mechanism now, thanks, I was not aware of it. It seems better the way you propose.

**Comment by ejgallego** [comment]
@herbelin I rebased to run CI again and be thus in the safe side, the delta was more than 200 commits.

**Pull request merged**
